docs/scraping-receivers.md

```diff
@@ -0,0 +1,143 @@
+between the steps to give users time to prepare, e.g. if the step one is released in v0.62.0, the step two should be 
+released in 0.65.0. Any warnings SHOULD include the version starting from which the next step will take action. If 
+breaking change is more complicated and many metrics are involved in the change, feature gates SHOULD be used instead.
+
+### Removing an optional metric
+
+Steps to remove an optional metric:
+
+1. Mark the metric as deprecated in `metadata.yaml` by adding "[DEPRECATED]" in its description. Show a warning that 
+   the metric will be removed if the `enabled` option is set explicitly to `true` in user settings. 
+2. Remove the metric.
+
+### Removing a default metric
+
+Steps to remove a default metric:
+
+1. Mark the metric as deprecated in `metadata.yaml` by adding "[DEPRECATED]" in its description. Show a warning that
+   the metric will be removed if the `enabled` option is not set explicitly to `false` in user settings.
+2. Turn the metric into optional. Show a warning that the metric will be removed if the `enabled` option is set 
+   explicitly to `true` in user settings.
+3. Remove the metric.
+
+### Replacing an optional metric
+
+Steps to replace an optional metric with another optional metric:
+
+1. Mark the metric as deprecated in `metadata.yaml` by adding "[DEPRECATED]" in its description. Add a warning that 
+   the metric will be replaced if the `enabled` option is set explicitly to `true` in user settings. Add the new
+   optional metric in `metadata.yaml`. 
+2. Remove the old metric.
+
+### Replacing a default metric
+
+Steps to replace a default metric with another default metric:
+
+1. Mark the metric as deprecated in `metadata.yaml` by adding "[DEPRECATED]" in its description. Show a warning 
+   that the metric will be replaced if the `enabled` option is not set explicitly to `false` in user settings. Add the
+   new metric as optional for now. Show another warning for the new metric if the `enabled` option is not set 
+   explicitly to any value in user settings.
+2. Turn the old metric into optional. Turn the new metric into default. Show the warning in case if user keeps the 
+   old metric enabled explicitly.
+3. Remove the old metric.
+
+### Making a default metric optional
+
+Steps to change metric from default to optional:
+
+1. Add a warning that the metric will be turned into optional if `enabled` field is not set explicitly to any value in
+   user settings. Warning example: "WARNING: Metric `foo.bar` will be disabled by default in v0.65.0. If you want to
+   keep it, please enable if explicitly in the receiver settings."
+2. Remove the warning and update `metadata.yaml` to make the metric optional.
+
+### Adding a new default metric or turning an existing optional metric into default
+
+Adding a new default metric is a breaking for a scraping receiver because it introduces an unexpected output for users 
+and additional load on metric backends. Steps to apply such change:
```

Comment:
I'm not sure this is really necessary or even reasonable.

With any other signal type, you would expect to get extra data that you do not care about, and either filter it or just ignore it.

Over time metrics will be added for many reasons and our scrapers will become less and less user friendly. 

To separate the concerns listed here:
- Are there backends that are actually "broken" by unexpected metrics?
- As long as additional load is roughly linear, is this really a problem?

Reply:
> Are there backends that are actually "broken" by unexpected metrics?
> As long as additional load is roughly linear, is this really a problem?

Not "broken" but can add an unexpected increase in billing at least. For example, in Splunk Infrastructure Monitoring, some metrics are categorized by use case and billed by number of identified entities, e.g. number of hosts or containers. Each category has a predefined set of metrics. Other metrics are going to a "custom" category which is billed by a number of time series. If we don't consider it as a breaking change and don't notify users about a new metric that will be enabled by default, they will get an unexpected increase in the custom time series affecting their billing.

Reply:
> Over time metrics will be added for many reasons and our scrapers will become less and less user friendly.

Not sure I understand why scrapers will become less and less user friendly?

Reply:
> Not sure I understand why scrapers will become less and less user friendly?

Let's say we enable N metrics for a scraper at the time it is marked stable, over time this set of metrics will become outdated, often through events that are outside of our control:
- Source technologies will deprecate and remove metrics from their APIs
- Source technologies will add new metrics, some of which are critical
- We will realize that we missed important metrics
- We will occasionally make a mistake and realize that a metric we added is problematic (e.g. causes excessive load on the source)

Over time, the set of metrics that we launched with grows stale, and in order for users to find the scraper useful, they will have to make more and more modifications to the default set of metrics. 

I think it would be unfortunately if we are unable to make sensible adjustments to the scraper's output from time to time. In my opinion, metrics should each have their own lifecycle so that our scrapers are not frozen in the past forever. This model would reflect the reality of the situation much better. We should look for another way to define and control a static _set_ of metrics.

Reply:
I completely agree that the set of default metrics should be always up to date and we shouldn't freeze them when we mark the component stable. That's why I mentioned that adding/removing metrics is possible in the stable state, but I don't think we should promote doing it silently. What if we relieve the requirement for such changes in the beta/stable states and mention all the acceptable reasons that you provided? Or you don't agree that we should call them "breaking" at all?

Reply:
> That's why I mentioned that adding/removing metrics is possible in the stable state, but I don't think we should promote doing it silently. What if we relieve the requirement for such changes in the beta/stable states and mention all the acceptable reasons that you provided?

Ok, I agree with what you're suggesting here. We can keep the metrics up to date by _expecting_ breaking changes, while also communicating them in a way that is easily noticed.

Reply:
Updated the definition of stable to expect breaking changes in order to keep the metrics set up to date.